compiler/rustc_lint/src/lints.rs

```diff
@@ -59,7 +59,18 @@ pub(crate) enum ShadowedIntoIterDiagSub {
 #[derive(LintDiagnostic)]
 #[diag(lint_implicit_unsafe_autorefs)]
 #[note]
-pub(crate) struct ImplicitUnsafeAutorefsDiag {
+pub(crate) struct ImplicitUnsafeAutorefsDiag<'a> {
+    #[label(lint_raw_ptr)]
+    pub raw_ptr_span: Span,
+    pub raw_ptr_ty: Ty<'a>,
+    #[note(lint_autoref)]
+    pub autoref_span: Span,
+    pub autoref_ty: Ty<'a>,
+    #[note(lint_method_def)]
+    pub method_def_span: Option<Span>,
+    pub method_name: Symbol,
+    #[note(lint_through_overloaded_deref)]
```

Comment:
Though what magic does the name `lint_through_overloaded_deref` resolve to `lint_implicit_unsafe_autorefs.through_overloaded_deref` in the flt file...?

Reply:
The `#[diag]` of the `#[derive(LintDiagnostic)]` defines the FTL message to use.
https://github.com/rust-lang/rust/blob/e4fac05277e9b98fc051ba25d857bd240ec66385/compiler/rustc_lint/src/lints.rs#L60
Here `lint_implicit_unsafe_autorefs`. *(Note: every `#[note]`, `#[label]`, `#[suggestion]` will be under that message.)*

Then we take the "custom name" `lint_through_overloaded_deref` strip the `rustc` crate name, here `lint`, which let's us with `through_overloaded_deref`, which combined with the diag resolves to `lint_implicit_unsafe_autorefs.through_overloaded_deref` in the FTL file.

We have some documentation about that at https://rustc-dev-guide.rust-lang.org/diagnostics/translation.html

Reply:
Thanks!

> strip the rustc crate name

Ah, that's the part I missed. Magic... I think with less magic this would be much easier to understand.^^

I can't find that particular quirk mentioned in the docs, either.

Reply:
In fact, the example used there does *not* put the crate prefix in the name for field attributes. If that also works, that seems like the better style to me -- avoiding relying on magic?

Reply:
The documentation is out-of-date, all slugs must be prefixed with the crate name.

```
error[E0425]: cannot find value `autoref` in module `crate::fluent_generated`
  --> compiler/rustc_lint/src/lints.rs:66:12
   |
66 |     #[note(autoref)]
   |            ^^^^^^^ not found in `crate::fluent_generated
```

Even in the documentation they use `hir_analysis_previous_decl_label` in the impl of `Diagnostic`.

Reply:
But if the prefix is stripped anyway then what sense does it make to require the user to write it? All it seems to do is make it harder to understand how the name is looked up...

Reply:
I think technically it's the inverse, I think the prefix is added by our `fluent_messages` macro when creating the values.

https://github.com/rust-lang/rust/blob/19738720130a82959acd4fc45259166262f3ffbe/compiler/rustc_lint/src/lib.rs#L139

As for why it's done like this, I don't know, I don't have the historical knowledge; and the current system is "pending a redesign" anyway.

Reply:
Oh I see...

The current system is "dead in the water", causing confusion and extra work with all its downsides but with no consensus to remove it and nobody attempting to improve it. :/